storage-provider/server/src/main.rs

```diff
@@ -390,63 +367,77 @@ impl Server {
-            p2p_task,
-            indexer_task,
+        tasks.spawn(
+            p2p::Worker::new(p2p_args)?
+                .run(cancellation_token.child_token())
+                .map(|result| ("P2P", result.map_err(ServerError::from))),
         );
+        tasks.spawn(
+            start_upload_server(Arc::new(storage_state), cancellation_token.child_token())
+                .map(|result| ("HTTP Storage", result.map_err(ServerError::from))),
+        );
+        tasks.spawn(
+            start_pipeline(
+                Arc::new(pipeline_state),
+                pipeline_rx,
+                cancellation_token.child_token(),
+            )
+            .map(|result| ("Pipeline", result.map_err(ServerError::from))),
+        );
+        tasks.spawn(
+            start_indexer(indexer_state, indexer_rx, cancellation_token.child_token())
+                .map(|result| ("Indexer", result.map_err(ServerError::from))),
+        );
+        tracing::info!("Successfully launched all sub-services, ready for work!");
 
-        // Exit with error
-        upload_result??;
-        rpc_task??;
-        pipeline_task??;
-        p2p_task??;
-        indexer_task??;
-
-        Ok(())
+        // Keep the first error around as the "canonical return",
+        // since we can't return multiple values
+        let mut error = None;
+        loop {
+            tokio::select! {
+                result = tasks.join_next() => {
+                    match result {
+                        Some(Ok((task_name, task_result))) => {
+                            match task_result {
+                                Ok(()) => tracing::info!("{task_name} finished successfully!"),
+                                Err(err) => {
+                                    tracing::error!("{task_name} finished with error: {err}");
+                                    tracing::error!("Cancelling remaining tasks...");
+                                    if error.is_none() {
+                                        error = Some(err);
+                                    }
+                                    cancellation_token.cancel();
+                                },
+                            }
+                        }
+                        Some(Err(err)) => {
+                            tracing::error!("Failed to join task with error: {err}");
+                            cancellation_token.cancel();
+                        }
+                        None => {
+                            // This branch should run when all tasks have terminated,
+                            // thus, this is the most appropriate place to return the value
+                            match error {
+                                Some(err) => return Err(err),
+                                None => return Ok(()),
+                            }
+                        },
+                    }
+                }
+                result = tokio::signal::ctrl_c() => {
```

Comment:
I think I've added the original impl and it's wrong. The ctrl_c is listening for the SIGINT signal not SIGTERM. For the more future proof signal handling the solution below should be better. It captures SIGINT and SIGTERM. It's good to capture the SIGTERM too, so that the process is also gracefully shutdown when running in a container.
 
```suggestion
/// Wait for a shutdown signal.
async fn wait_for_shutdown_signal() {
    use tokio::signal::unix::{signal, SignalKind};

    // Infos here:
    // https://www.gnu.org/software/libc/manual/html_node/Termination-Signals.html
    let mut signal_terminate = signal(SignalKind::terminate()).expect("should be able to listen for a SIGTERM");
    let mut signal_interrupt = signal(SignalKind::interrupt()).expect("should be able to listen for a SIGINT");

    tokio::select! {
        _ = signal_terminate.recv() => tracing::debug!("Received SIGTERM."),
        _ = signal_interrupt.recv() => tracing::debug!("Received SIGINT."),
    };
}
```